crates/primitives/src/chain/spec.rs

```diff
@@ -491,7 +501,16 @@ impl ChainSpec {
             if let ForkCondition::Block(block) |
             ForkCondition::TTD { fork_block: Some(block), .. } = cond
             {
-                if cond.active_at_head(head) {
+                // the timestamp check here is a bit brittle as its tied to the self.satisfy() impl
+                // returning a non-0 head.timestamp for ForkCondition::Timestamp, and a 0
+                // timestamp for the other ForkCondition types
+                // this check is required though - without it, we will drop into the else block and
+                // shortcircuit for heads valid during timestamp-based forks, leading to an
+                // incorrect ForkId for Shanghai, Cancun etc. An alternative fix
+                // here would be to modify the Head returned in self.satisfy() to
+                // include a valid blocknum for ForkCondition::Timestamp, thus not
+                // excercising the else block, and calculating an incorrect fork_hash
```

Comment:
I really don't understand why we need this now,

this PR was supposed to add support for this on Chainspec

https://github.com/paradigmxyz/reth/blob/b7a6eedd3a92f0105d26b255d1caed0a5b0e0eee/crates/primitives/src/hardfork.rs#L47-L53

Reply:
`spec.fork_id(&spec.fork(*self).satisfy())` this line has a bug for `Hardfork::Shanghai` and all timestamp based HardForks - I have removed this block and push up the assoc. test failures.

I caught it while adding unit tests - The existing test coverage differs from the `satisfy()` actual implementation, and mine hooks into the underlying `satisfy()` call.


See:
```
running 1 test
thread 'chain::spec::tests::mainnet_hardfork_fork_ids' panicked at 'assertion failed: `(left == right)`
  left: `ForkId { hash: ForkHash("dce96c2d"), next: 0 }`,
 right: `ForkId { hash: ForkHash("fc64ec04"), next: 1150000 }`: Expected fork ID ForkId { hash: ForkHash("dce96c2d"), next: 0 }, computed fork ID ForkId { hash: ForkHash("fc64ec04"), next: 1150000 } for hardfork Shanghai', crates/primitives/src/chain/spec.rs:1179:17
stack backtrace:
```

Happy to sync with @Rjected on this further.

Reply:
Without my change the `ForkId` for Shanghai and presumably all the future timestamp based forks (Cancun etc) will short-circuit the existing logic and return the `ForkId` for Frontier

Reply:
This is probably a bug in `satisfy`, since to obtain the `ForkId` for shanghai, the `fork_id` method uses the block number of the `Head` argument to apply the block-based forks, before adding the timestamp-based forks. Satisfy does this:
https://github.com/paradigmxyz/reth/blob/7a0781c25e2b46e7207f90160eed8ea4b6b591ec/crates/primitives/src/chain/spec.rs#L939-L943

`fork_id` does this:
https://github.com/paradigmxyz/reth/blob/7a0781c25e2b46e7207f90160eed8ea4b6b591ec/crates/primitives/src/chain/spec.rs#L488-L498

So for `satisfy` to work with timestamp forks, it needs to return a `Head` with a block and TTD greater than the highest known fork block / TTD respectively. It should also not use `Default::default`.

Reply:
yeah exactly!

called similar out in the comments:
> An alternative fix
                // here would be to modify the Head returned in self.satisfy() to
                // include a valid blocknum for ForkCondition::Timestamp, 
                
I can fix it there instead of this timestamp fix, I'll push that up in the next rev

Reply:
ah so... I spent a bit of time looking into this, and fixing it in `satisfy()` is a little bit more difficult than I thought.

That method is type `ForkCondition::satisfy()` and effectively only has access each enum variant's values. So in this case. `satisfy()` only has access to either 
```
ForkCondition::Block(block_num) => Head { block_num, .. Default())
ForkCondition::Timestamp(timestamp) = Head {timestamp, .. Default())
```

so if we want to fix it in `satisfy()` we may need to change `ForkCondition::Timestamp(u64,u64)` to include both the `block_num` and `timestamp` - and change it everywhere its been declared/initalized.

It will be a larger code change - which I'm open to take on. The alternative to fixing it in `satisfy()` is to instead just add 

```
if cond.active_at_head(head) || head.timestamp > 0 
```
like it was in the first push up the prev commit to this PR. 

Wdyt @Rjected - should we refactor `ForkCondition::Timestamp` to include a block_num in the variant vals? It feels like this may take away from the whole "timestamp based fork condition" aspect, but it may also be necessary. 

Happy to implement another solution as well - if one exists.

Reply:
also - it doesn't make sense to just hardcode a blocknum/ttd in that `Head{} ` return because it varies per chain.

so mainnet's forkcondition::timestamp for Shanghai/Cancun will be different than sepolias.

Reply:
I don't really want to add in block nums to timestamp forks. Maybe `ForkCondition` is the wrong place to put satisfy though - looking at the two places where it's used:
https://github.com/paradigmxyz/reth/blob/4603dc31c03747119878800216f87d2f6a3635f4/crates/primitives/src/hardfork.rs#L47-L62

It looks like we have access to the `ChainSpec` - so this could be `ChainSpec::satisfy` instead of `ForkCondition::satisfy`

Reply:
alright, let me take a look.

Reply:
yeah.. hmm I'm a little unclear still on how moving it Chainspec would help here, are you suggesting we do a match/check against the spec within the `satisfy()` method to return a valid Head?

That may work for Mainnet, Sepolia etc - but IIUC, ChainSpec is also used to create custom chainspecs that aren't bound to an actual chain

Would love some help thinking through the impl